packages/grid/src/vaadin-grid-array-data-provider-mixin.js

```diff
@@ -59,11 +62,7 @@ export const ArrayDataProviderMixin = (superClass) =>
             size: 0,
           });
           this.clearCache();
-        } else if (this._arrayDataProvider.__items === items) {
-          // The items array was modified
-          this.clearCache();
```

Comment:
I had to move this branch into a separate observer as it was causing some tests to fail.

The branch was invoked on data provider change. However, it was always invoked before the `dataProviderChanged` observer which updates the controller's data provider callback, so `clearCache()` would load data with the previous data provider.